### PR TITLE
Implemented signup form

### PR DESCRIPTION
**Pull request by Himanshusingh770**

### CODE REVIEW

File: src/Components/SucessModel.jsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  const handleOkClick = () => {
+    navigate('/');
+    onClose(); 
+  };
+
+  return (
+    <Modal show={show} centered size="sm">
+    <Modal.Body className="text-center">
+      <p className="text-success fs-4">Sign Up Successfully!!</p>
+      <p>Your account has been created successfully. You can now log in.</p>
```
Comment (PiyushMangla) [inline src/Components/SucessModel.jsx]: use props for message

File: src/Components/SucessModel.jsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import { Modal, Button } from 'react-bootstrap';
+import { useNavigate } from 'react-router-dom';
+import '../App.css'
+const SuccessModal = ({ show, onClose }) => {
+  const navigate = useNavigate();
+  const handleOkClick = () => {
```
Comment (PiyushMangla) [inline src/Components/SucessModel.jsx]: use prop for ok click

File: src/Components/SucessModel.jsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      <p className="text-success fs-4">Sign Up Successfully!!</p>
+      <p>Your account has been created successfully. You can now log in.</p>
+    </Modal.Body>
+    <Modal.Footer className="justify-content-center ">
+      <Button
+        className="rounded-pill  ok-button"
+        variant="success"
+        onClick={handleOkClick}
+      >
+        OK
```
Comment (PiyushMangla) [inline src/Components/SucessModel.jsx]: use prop for button text

File: src/Pages/SignUpForm.jsx [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    lastName: '',
+    email: '',
+    roleId: '',    // Changed from role to roleId
+    password: '',
+    confirmPassword: ''
+  });
+  const [formErrors, setFormErrors] = useState({});
+  const [isLoading, setIsLoading] = useState(false);
+  const [showToast, setShowToast] = useState(false);
+  const [toastMessage, setToastMessage] = useState(''); // for internal server 500 error
```
Comment (PiyushMangla) [inline src/Pages/SignUpForm.jsx]: use object for toast data all 3in one

File: src/Pages/SignUpForm.jsx [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+  const [toastVariant, setToastVariant] = useState('success'); // Variant state
+  const [showPassword, setShowPassword] = useState(false);
+  const [showConfirmPassword, setshowConfirmPassword] = useState(false);
+  const [showModal, setShowModal] = useState(false);
```
Comment (PiyushMangla) [inline src/Pages/SignUpForm.jsx]: which modal change name

File: src/Pages/SignUpForm.jsx [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    roleId: '',    // Changed from role to roleId
+    password: '',
+    confirmPassword: ''
+  });
+  const [formErrors, setFormErrors] = useState({});
+  const [isLoading, setIsLoading] = useState(false);
+  const [showToast, setShowToast] = useState(false);
+  const [toastMessage, setToastMessage] = useState(''); // for internal server 500 error
+  const [toastVariant, setToastVariant] = useState('success'); // Variant state
+  const [showPassword, setShowPassword] = useState(false);
```
Comment (PiyushMangla) [inline src/Pages/SignUpForm.jsx]: use object for these two too password and confirm password

File: src/Pages/SignUpForm.jsx [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    const { errors, isFormValid } = validateForm(formData, true);
+    setFormErrors(errors);
+
+    if (!isFormValid) return;
+
+    setIsLoading(true);
+    dispatch(signup(formData))
+      .unwrap()
```
Comment (PiyushMangla) [inline src/Pages/SignUpForm.jsx]: use try catch

File: src/Pages/SignUpForm.jsx [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+          // Fallback message for cases where there's no error message from the API
+          setToastMessage('Server Not responed.');
+          setToastVariant('danger');
+        }
+        setShowToast(true);
+        
+      });
+  };
+  const togglePasswordVisibility = () => setShowPassword(!showPassword);
+  const toggleConfirmPasswordVisibility = () =>
```
Comment (PiyushMangla) [inline src/Pages/SignUpForm.jsx]: create single function for both toggle

File: src/Pages/SignUpForm.jsx [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+                  <option disabled>No roles available</option>
+                )}
+              </Form.Control>
+              <Form.Control.Feedback type="invalid">
+                {formErrors.roleId}
+              </Form.Control.Feedback>
+            </Form.Group>
+
+            <Form.Group controlId="formPassword" className="mb-1 px-4">
+              <Form.Label>Password</Form.Label>
```
Comment (PiyushMangla) [inline src/Pages/SignUpForm.jsx]: create a separate component for password field and use from there.
Its not looking nice

File: src/slices/authSlice.jsx [hunk 10]
```diff
@@ -1,52 +1,55 @@
...
   name: 'auth',
   initialState: {
-    user: null, // Initially, no user is stored
-    isAuthenticated: false, // User is not authenticated by default
+    user: null,
+    isAuthenticated: false,
     isLoading: false,
-    error: null
+    error: null,
+    roles: [],
```
Comment (PiyushMangla) [inline src/slices/authSlice.jsx]: why its here

File: src/Components/PasswordField.jsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
+// PasswordField.js
+import React from 'react';
+import { Form } from 'react-bootstrap';
+import { EyeFill, EyeSlashFill } from 'react-bootstrap-icons';
+
+const PasswordField = ({ label, name, value, onChange, error, showPassword, toggleVisibility }) => {
```
Comment (yuvrajInno) [inline src/Components/PasswordField.jsx]: showpassword , togglevisibility should not used be as props

File: src/Components/SucessModel.jsx [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { Modal, Button } from 'react-bootstrap';
+import '../App.css';
+
+const SuccessModal = ({ show, onClose, message, okButton, onOkClick }) => {
+  return (
+    <Modal show={show} centered size="sm">
+      <Modal.Body className="text-center">
+        <p className="text-success fs-4">{message}</p>
+        <p>Your account has been created successfully. You can now log in.</p>
```
Comment (yuvrajInno) [inline src/Components/SucessModel.jsx]: Why not prop?

File: src/Pages/SignUpForm.jsx [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+  const [formErrors, setFormErrors] = useState({});
+  const [isLoading, setIsLoading] = useState(false);
+  const [toast, setToast] = useState({
+    show: false,
+    message: '',
+    variant: 'success'
+  });
+  const [showPassword, setShowPassword] = useState(false);
+  const [showConfirmPassword, setShowConfirmPassword] = useState(false);
+  const [showOkModal, setshowOkModal] = useState(false);
```
Comment (yuvrajInno) [inline src/Pages/SignUpForm.jsx]: change variable name

File: src/Components/TostMassege.jsx [hunk 14]
```diff
@@ -1,24 +1,40 @@
-import React from 'react';
```
Comment (yuvrajInno) [inline src/Components/TostMassege.jsx:1]: correct component name

File: src/Components/SucessModel.jsx [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
```
Comment (yuvrajInno) [inline src/Components/SucessModel.jsx]: correct component name

File: src/Components/ButtonModel/CustomButtonModel.jsx [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
```
Comment (yuvrajInno) [inline src/Components/ButtonModel/CustomButtonModel.jsx:1]: change component name why it has model in it?

File: src/App.css [hunk 17]
```diff
@@ -29,20 +31,41 @@
...
+  .button-custom {
     display: flex;
     justify-content: center;
     align-items: center;
     margin-top: 1rem;
     width: 100%;
+    padding: 10px; 
+    border-radius: 8px; 
+    font-weight: bold; 
+    border: none;
```
Comment (yuvrajKaexpertise) [inline src/App.css:43]: you are using tailwind CSS, so why are there too many custom styles? @Himanshusingh770

File: src/Data/roles.json [hunk 18]
```diff
@@ -0,0 +1,31 @@
+[
```
Comment (yuvrajKaexpertise) [inline src/Data/roles.json]: why this static data? @Himanshusingh770

File: src/Pages/LoginForm.jsx [hunk 19]
```diff
@@ -2,12 +2,13 @@ import React, { useState } from 'react';
...
+import { Form, Container, Card } from 'react-bootstrap';
 import { EyeFill, EyeSlashFill } from 'react-bootstrap-icons';
 import { validateField, validateForm } from '../utils/validationUtils';
-import ToastMessage from '../Components/TostMassege';
-import "../App.css"
-
+import ToastMessage from '../Components/ToastMessage';
+import { Link } from 'react-router-dom';
+import CustomButton from '../Components/ButtonModal/CustomButtonModel';
+import '../App.css';
```
Comment (yuvrajKaexpertise) [inline src/Pages/LoginForm.jsx]: why app.css imported in LoginForm? @Himanshusingh770

File: src/Pages/SignUpForm.jsx [hunk 20]
```diff
@@ -0,0 +1,225 @@
+import React, { useState, useEffect } from 'react';
+import { useDispatch,useSelector } from 'react-redux';
+import { Link, useNavigate } from 'react-router-dom';
+import { Form, Container, Card } from 'react-bootstrap';
+import ToastMessage from '../Components/ToastMessage'; // Ensure the path is correct
+import { signup } from '../slices/authSlice';
+import { fetchRoles } from '../slices/rolesSlice';
+import { validateField, validateForm } from '../utils/validationUtils';
+import '../App.css';
```
Comment (yuvrajKaexpertise) [inline src/Pages/SignUpForm.jsx]: same here

File: src/Pages/SignUpForm.jsx [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+    variant: 'success'
+  });
+  const [SignupSuccessModal, setSignupSuccessModal] = useState(false);
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+  // const roles = useSelector((state) => state.roles.roles);
+
+  useEffect(() => {
+    // dispatch(fetchRoles()); // use API call
+  }, [dispatch]);
```
Comment (yuvrajKaexpertise) [inline src/Pages/SignUpForm.jsx:38]: blank use effect ? @Himanshusingh770

**Review by yuvrajKaexpertise — CHANGES_REQUESTED** [review]

**Pull request merged**